Design/ISIS_Imaging/ISIS_Imaging_general.md

```diff
@@ -0,0 +1,182 @@
+
+Current issues for this section:
+- Dynamic dialogue building
+  - Solution is to use an approach like the `cli_registrator`, but pass in the QObject in which the filters will register themselves in
+  - Dialogue registration is a bit more complicated, because we have to connect the dropdown menu to the dialogue's `.show()` method.
+- Transferring information from the dialogue (the parameters) to the execution
+  - Current solution is to use a partial function, decorated with all of the parameters. No other function down the chain of execution will have to worry about the parameters that way.
+- Handling filters with multiple steps
+  - Background correction and contrast normalisation need to calculate values for scaling, do the filter execution, and then apply the scaling to the images.
+     - Currently the solution is to apply a monad style approach.
+
+## Undoing an operation
+Building the Undo operation may be complicated, but here are a few high level approaches:
+- Storing the original state as a deepcopy, and then applying the filter inplace on the volume of images, works best with single image.
+    - Serious memory implications for _very_ large volumes.
+    - Should be decent for small stacks, or when loading images with a large step.
+    - Can easily store a single undo (like imagej), multiple undos can be hard to implement efficiently because of all the memory being copied
+        - Naive implementation is to append all the consecutive changes in a list.
+        - Or we can just not provide an undo if the operation was applied to the STACK.
+    - Multiple undos implementation is to store it as part of the History in the process list.
+        - Store slice index, operation(+params) and result(deepcopy) as an entry in a process list
+            - Pro: Can easily revert an image to any state (if we don't delete the newer states we can switch between them to see the difference)
+            - Cons: Memory and might be convoluted to handle the history
+            - Option to open the saved image in a new stack
+        - This means every slice in a volume has unique history.
+            - That might be something that will potentially cause problems as it is complicated ( and error prone? ). It needs to be well implemented and tested.
+            - But also provides the opportunity for the user to easily visualise different operations' results.
+
+## Histograms
+Computation of the histograms should be done only on the ROI, if no ROI is selected do the whole image. The computation itself should ideally take place in the `core.algorithms` package.
+
+Visualisation can be one of the two:
+- As a small widget inside the stack window
+- As a small independent widget floating around.
+
+A problem that needs to be considered is, should we allow for more than _one_ histogram to be active at a time? For example ImageJ does not.
```

Comment:
Is this a memory issue? e.g Is this because for a 16-bit image you have an array with 65536 entries? Otherwise you have to handle the histogram "window" being updated every time you select a new image so users know which image the histogram corresponds to.